# Understanding Your Data

### "Understanding the data comes before analysis"

In [1]:
# Load a dataset
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
data={
  "Student":["A","B","C","D","E","F","G","H"],
  "Age":[20,21,20,22,21,20,23,21],
  "Gender":["F","M","F","M","F","M","F","F"],
  "Study_Hours":[5,2,7,4,6,3,8,5],
  "Marks":[78,55,88,65,82,60,92,75],
  "Passed":["Yes","No","Yes","Yes","Yes","No","Yes","Yes"]
}

df=pd.DataFrame(data)
df

,Student,Age,Gender,Study_Hours,Marks,Passed
0,A,20,F,5,78,Yes
1,B,21,M,2,55,No
2,C,20,F,7,88,Yes
3,D,22,M,4,65,Yes
4,E,21,F,6,82,Yes
5,F,20,M,3,60,No
6,G,23,F,8,92,Yes
7,H,21,F,5,75,Yes


In [4]:
df.head()

,Student,Age,Gender,Study_Hours,Marks,Passed
0,A,20,F,5,78,Yes
1,B,21,M,2,55,No
2,C,20,F,7,88,Yes
3,D,22,M,4,65,Yes
4,E,21,F,6,82,Yes


In [5]:
df.tail()

,Student,Age,Gender,Study_Hours,Marks,Passed
3,D,22,M,4,65,Yes
4,E,21,F,6,82,Yes
5,F,20,M,3,60,No
6,G,23,F,8,92,Yes
7,H,21,F,5,75,Yes


In [6]:
df.shape

(8, 6)

In [8]:
# no.of columns to show in shape
df.shape[1]

6

In [9]:
df.columns

Index(['Student', 'Age', 'Gender', 'Study_Hours', 'Marks', 'Passed'], dtype='object')

In [10]:
df.columns.tolist()

['Student', 'Age', 'Gender', 'Study_Hours', 'Marks', 'Passed']

In [11]:
df.dtypes

,0
Student,object
Age,int64
Gender,object
Study_Hours,int64
Marks,int64
Passed,object


In [12]:
# info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Student      8 non-null      object
 1   Age          8 non-null      int64 
 2   Gender       8 non-null      object
 3   Study_Hours  8 non-null      int64 
 4   Marks        8 non-null      int64 
 5   Passed       8 non-null      object
dtypes: int64(3), object(3)
memory usage: 516.0+ bytes


In [13]:
# Missing values
df.isnull()

,Student,Age,Gender,Study_Hours,Marks,Passed
0,False,False,False,False,False,False
1,False,False,False,False,False,False
2,False,False,False,False,False,False
3,False,False,False,False,False,False
4,False,False,False,False,False,False
5,False,False,False,False,False,False
6,False,False,False,False,False,False
7,False,False,False,False,False,False


In [14]:
df.isnull().sum()

,0
Student,0
Age,0
Gender,0
Study_Hours,0
Marks,0
Passed,0


In [16]:
# Percentage of missing values
missing_percentage=df.isnull().mean()*100
missing_percentage

,0
Student,0.0
Age,0.0
Gender,0.0
Study_Hours,0.0
Marks,0.0
Passed,0.0


In [17]:
# Duplicate rows
df.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
5,False
6,False
7,False


In [18]:
df.duplicated().sum()

np.int64(0)

In [19]:
# Remove duplicates
df=df.drop_duplicates()
df

,Student,Age,Gender,Study_Hours,Marks,Passed
0,A,20,F,5,78,Yes
1,B,21,M,2,55,No
2,C,20,F,7,88,Yes
3,D,22,M,4,65,Yes
4,E,21,F,6,82,Yes
5,F,20,M,3,60,No
6,G,23,F,8,92,Yes
7,H,21,F,5,75,Yes


In [20]:
# Numerical summary
df.describe()

,Age,Study_Hours,Marks
count,8.000000,8.00,8.000000
mean,21.000000,5.00,74.375000
std,1.069045,2.00,13.298093
min,20.000000,2.00,55.000000
25%,20.000000,3.75,63.750000
50%,21.000000,5.00,76.500000
75%,21.250000,6.25,83.500000
max,23.000000,8.00,92.000000


In [21]:
# Categorical summary
df["Gender"].describe()

,Gender
count,8
unique,2
top,F
freq,5


In [25]:
# Unique values
df['Gender'].unique()

array(['F', 'M'], dtype=object)

In [26]:
df["Gender"].nunique()

2

In [27]:
# Value counts
df['Gender'].value_counts()

,count
Gender,
F,5
M,3


In [28]:
# Value count percentage
df["Gender"].value_counts(normalize=True) * 100

,proportion
Gender,
F,62.5
M,37.5


In [29]:
# Checking suspicious values

## This is where real data cleaning begins.
df["Gender"].unique()

array(['F', 'M'], dtype=object)

In [31]:
# Checking numerical ranges
df["Age"].describe()

,Age
count,8.000000
mean,21.000000
std,1.069045
min,20.000000
25%,20.000000
50%,21.000000
75%,21.250000
max,23.000000


In [32]:
# nunique() for every column
df.nunique()

,0
Student,8
Age,4
Gender,2
Study_Hours,7
Marks,8
Passed,2


## Key Findings:

1. The dataset contains 8 rows and 6 columns.

2. The numerical columns are 3.

3. The categorical columns are 2.

4. The dataset contains 0 missing values.

5. The dataset contains 0 duplicate rows.

6. The average marks are 74.375.

7. The median marks are 76.5.

8. The minimum marks are 55 and the maximum marks are 92.

9. There are 2 unique Gender categories.

10. The most frequent Gender category is Female(F).

11. One potential data-quality issue I noticed is -
(1)- the passing threshold is not specified, so we cannot verify whether the Passed values are correctly assigned..
(2)-No information about the units/context of Study_Hours
Are these hours per day, per week, or total study hours? The column name alone doesn't tell us.
(3)-No information about the units/context of Study_Hours
Are these hours per day, per week, or total study hours? The column name alone doesn't tell us.

12. Before performing EDA, I should first understand the data.